File: schedule-service/src/main/java/mrsisk/github/io/appointservice/service/TokenService.java
package mrsisk.github.io.appointservice.service;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.security.Key;
import java.util.Date;
import java.util.Map;
import java.util.function.Function;

@Service
public class TokenService {

    @Value("${jwt.secret}")
    String secret;

    public static final long TOKEN_VALIDITY = 60 * 60 * 24;

    String doGenerateToken(Map<String, Object> claims, String subject) {
        return Jwts.builder()
                .setClaims(claims)
                .setSubject(subject)
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + TOKEN_VALIDITY * 1000))
                .signWith(getSignInKey(), SignatureAlgorithm.HS256)
                .compact();
    }
    private Key getSignInKey() {
        byte[] keyBytes = Decoders.BASE64.decode(secret);
        return Keys.hmacShaKeyFor(keyBytes);
    }

    public String getClaim(String token, String key){
        Claims claim = getAllClaimsFromToken(token);
        return claim.get(key).toString();
    }

    public String getSubjectFromToken(String token) {
        return getClaimFromToken(token, Claims::getSubject);
    }

    //retrieve expiration date from jwt access_token
    public Date getExpirationDateFromToken(String token) {
        return getClaimFromToken(token, Claims::getExpiration);
    }


    private  <T> T getClaimFromToken(String token, Function<Claims, T> claimsResolver) {
        final Claims claims = getAllClaimsFromToken(token);
        return claimsResolver.apply(claims);
    }
    private Claims getAllClaimsFromToken(String token) {
        return Jwts.parserBuilder().setSigningKey(getSignInKey()).build().parseClaimsJws(token).getBody();
    }

    public boolean isTokenValid(String token){
        try {
            getAllClaimsFromToken(token);
            return true;
        }catch (Exception ex){
            return false;
        }
    }
}

File: accounts-service/src/main/kotlin/mrsisk/github/io/accountsmanager/service/TokenService.kt
package mrsisk.github.io.accountsmanager.service

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.io.Serial
import java.io.Serializable
import java.security.Key
import java.util.*
import java.util.function.Function

@Service
class TokenService:  Serializable {

    @Value("\${accounts-manager.security.token-secret}")
    private lateinit var secret: String


    @Value("\${accounts-manager.security.token-validity}")
    private lateinit var validity: String

    @Serial
    private val serialVersionUID = 234234523523L

    fun generateToken(subject: String, claims: Map<String, Any>): String{
        return Jwts.builder()
            .setClaims(claims)
            .setIssuer("task-manager")
            .setSubject(subject)
            .setIssuedAt(Date(System.currentTimeMillis()))
            .setExpiration(Date(System.currentTimeMillis() + tokenValidity()))
            .signWith(getSignInKey())
            .compact()
    }


    fun getClaim(key: String, token: String): Any? {
       val claims = getAllClaimsFromToken(token)
        return claims[key]
    }

    fun <T> getClaimFromToken(token: String, claimsResolver: Function<Claims, T>): T {
        val claims: Claims = getAllClaimsFromToken(token)
        return claimsResolver.apply(claims)
    }

    fun getAllClaimsFromToken(token: String): Claims {
        return Jwts.parserBuilder()
            .setSigningKey(getSignInKey())
            .build()
            .parseClaimsJws(token)
            .body
    }

    private fun getSignInKey(): Key {
        val keyBytes = Decoders.BASE64.decode(secret)
        return Keys.hmacShaKeyFor(keyBytes)
    }

    private fun tokenValidity(): Long{
        return try {
            validity.toLong()

        }catch (ex: Exception){
            86400000
        }
    }

}